### ISSUE TITLE
Add ability to print the article

### ISSUE THREAD

**Issue by Jaifroid** [kiwix/kiwix-js#350]
The work on this has been done in https://github.com/kiwix/kiwix-js-windows/issues/13 and is experimentally implemented in the latest versions of WikiMed and Wikivoyage in the Microsoft Store.

The issue is that in most browsers (all?) when invoking print in Kiwix JS, visible areas of the app interface are sent to the printer, and only one page is printed, even if the article in the iframe consists of multiple pages. The solution to this is to invoke `window.print()` inside the scope of the iframe. This can be done with `window.frames[0].frameElement.contentWindow.print();` and probably also with other DOM methods of getting the contentWindow of the iframe.

At the most basic level, what is needed is to intercept Ctrl-P both in the iframe and in the top-level document (it can fire in either context depending on where the user has clicked). However, most naïve users don't know about Ctrl-P, or don't use it, or may be using a Mac (though Command-P may be the equivalent there). So for them, a Print button or Print menu item has to be provided.

Finally, Wikipedia documents contain a number of elements that the user may not wish to print, e.g. navigation boxes that are useless in a printed document, or long lists of references that take up multiple pages. We can either offer the user a choice of elements to filter out (using CSS `@media print` rules), or we can just eliminate the most obvious ones behind the scenes (navigation elements).

In Kiwix JS Windows, I implemented a dialogue box with bootstrap modal to allow the user to include or exclude elements. Most of the elements are specific to Wikipedia and Wikivoyage.

I'll start by backporting the keydown intercept (it has to be keydown, or else the browser's own print dialogue opens anyway) and getting printing working from the iframe. Then we can decide what else we want.

**Comment by Jaifroid** [comment]
In #888 @mossroy comented:

> If you just use ctrl-p in your browser, it will (badly) print the UI and only the beginning of the content (see #887) It works well if you ask the browser to print the content of the iframe. On Firefox, you can right-click in the content, choose "this frame" -> "print the frame".
> 
> To avoid this issue, I don't see a better solution than to provide a "print" feature in the app, that would trigger the browser print on the iframe:
> 
> `document.getElementById("articleContent").contentWindow.print()`

**Comment by Jaifroid** [comment]
@BBentleyDev Take a look at the implementation of print on https://pwa.kiwix.org . It intercepts key presses for Ctrl-P and also has a button in the task bar for printing. It includes a dialogue box that allows the user to choose what to print for Wikipedia articles, but this might need to be made more generic, as we now support many more ZIM types, and it is probably not a good idea to tie the feature to any specific ZIM type.

The use of a print button in the UI may be dependent on #523.

**Comment by mossroy** [comment]
I confirm we should not do anything specific to a ZIM type here.
It probably means something much simpler: no printOptions, no settings in the UI.
But trapping ctrl-p is probably a good idea, I agree.

**Comment by Jaifroid** [comment]
It might be worth checking whether overriding the `window.print()` function so that it runs code to print the contents of the iframe instead of the main window, could provide a very simple solution that would work with the browser-provided print options (i.e. Ctrl-P and any UI printing controls). This would only work if the browser uses JavaScript for its own implementation of these controls, and this may vary per-browser. It's probably a long shot, but worth checking.

If above doesn't work, then as I wrote in the first post, it's necessary to trap the `keydown` event for Ctrl-P/Command-P, because `keyup` or `keypress` events don't fire before the browser intercepts Ctrl-P.

Finally, if it's not possible to intercept the browser's Print menu/controls, then I think a print button would be necessary, because you can't rely on any users knowing the keyboard shortcut -- only a (small) minority of users ever use keyboard shortcuts in my experience. In Kiwix JS Windows, I re-used the "About" button for this: it becomes a Print button when the user is viewing an article.

**Comment by mossroy** [comment]
I totally agree.
Except that I find questionable to re-use the About icon.
IIRC, it was to save space on devices with small screens?

If necessary, we could do that in two steps:

- add a "print" icon, next to the "random" one in this issue
- postpone to #523 the way to improve the UI

**Comment by Jaifroid** [comment]
Yes, it was too save screen space. I agree with the suggestion of postponing UI-related issues to #523.

**Comment by Rbcoder1** [comment]
I am currently working on #523 and just changed the navigation UI 

So still add the button for printing the content of zim file
And add it into navbar.

**Comment by Jaifroid** [comment]
The thing is, we don't have a print function in Kiwix JS. This issue would need to be implemented before adding a print button to the new UI.

**Comment by Rbcoder1** [comment]
Yes so i am just fix the ui of navigation bar and then add a print function it.

**Comment by Bam92** [comment]
Is this feature still needed?

**Comment by Jaifroid** [comment]
@Bam92 Yes, we still don't have a print feature in KJS. There is such a feature in https://pwa.kiwix.org. To fix this issue, it would be necessary to add a function like `printArticle()` (maybe to `uiUtil.js`) and evoke it when the user presses Ctrl-P / Command-P. Adding any UI for this would be a separate issue.

**Comment by Bam92** [comment]
I'm working on this.

**Comment by Adedejiosvaldo** [comment]
@Bam92 How is the progress made so far or would you recommend I get started @Jaifroid ?

**Comment by Adedejiosvaldo** [comment]
@THEBOSS0369 Yes I am currently working on it

**Comment by PARTH-TUSSLE** [comment]
Hi! I'd love to work on this issue. Could you please assign it to me? @mossroy @mossroy @Bam92 @Adedejiosvaldo @Rbcoder1

**Comment by devhenode** [comment]
Hello @Jaifroid I decided to try out this feature on my repo and I was able to see some cool changes (linking the image below).

But I am still experiencing some pages content not being able to printed while some zim files content get printed.

Anyone making contribution to this and experiencing the same issue?

[testing-testing-workinng-wikipage.pdf](https://github.com/user-attachments/files/19445525/testing-testing-workinng-wikipage.pdf)
[testing for bitcoin.pdf](https://github.com/user-attachments/files/19445526/testing.for.bitcoin.pdf)
[testing-testing-working-100.pdf](https://github.com/user-attachments/files/19445524/testing-testing-working-100.pdf)

![Image](https://github.com/user-attachments/assets/eb6fc46e-e77b-47af-a98b-fd5479262528)

**Comment by devhenode** [comment]
And I already made an improvement to the page UI to include a print icon button for it can be easier to print the content of the page . 
I had like you to view it in a more detailed view below. Let me know what anyone thinks of it. And if i should open a PR to get it approved

https://github.com/user-attachments/assets/2711df31-3265-40c8-b917-ee8a78d9b80e

**Comment by Prasannacharya-3151** [comment]
Hi, I am @Prasannacharya-3151 interested in working on this issue. Is it still available or should I pick another one??
